src/include/OSL/oslconfig.h

```diff
@@ -75,34 +152,55 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
     #endif
 #endif // OSL_DEBUG
 
-#ifndef OSL_INLINE
-    #if OSL_DEBUG
-        #define OSL_INLINE inline
-    #else
-        #if __INTEL_COMPILER >= 1100 || (defined(_WIN32) || defined(_WIN64))
-            #define OSL_INLINE __forceinline
-        #else
-            #define OSL_INLINE inline __attribute__((always_inline))
-        #endif
-    #endif
+
+// OSL_FORCEINLINE is a function attribute that attempts to make the
+// function always inline. On many compilers regular 'inline' is only
+// advisory. Put this attribute before the function return type, just like
+// you would use 'inline'. Note that if OSL_DEBUG is true, it just becomes
+// ordinary inline.
+#if OSL_DEBUG
+#    define OSL_FORCEINLINE inline
+#elif defined(__CUDACC__)
+#    define OSL_FORCEINLINE __inline__
+#elif defined(__GNUC__) || defined(__clang__) || __has_attribute(always_inline)
+#    define OSL_FORCEINLINE inline __attribute__((always_inline))
+#elif defined(_MSC_VER) || defined(__INTEL_COMPILER)
+#    define OSL_FORCEINLINE __forceinline
+#else
+#    define OSL_FORCEINLINE inline
 #endif
 
-#ifndef OSL_NOINLINE
-    #if (defined(_WIN32) || defined(_WIN64))
-        #define OSL_NOINLINE __declspec(noinline)
-    #else
-        #define OSL_NOINLINE __attribute__((noinline))
-    #endif
+
+// OSL_NOINLINE hints to the compiler that the functions should never be
+// inlined.
+#if defined(__GNUC__) || defined(__clang__) || __has_attribute(noinline)
+#    define OSL_NOINLINE __attribute__((noinline))
+#elif defined(_MSC_VER) || defined(__INTEL_COMPILER)
+#    define OSL_NOINLINE __declspec(noinline)
+#else
+#    define OSL_NOINLINE
 #endif
 
-#if defined(_MSC_VER) && !defined(__INTEL_COMPILER)
-    #define OSL_EXPECT_FALSE(EXPRESSION) EXPRESSION
-    #define OSL_EXPECT_TRUE(EXPRESSION) EXPRESSION
+
+// Some compilers define a special intrinsic to use in conditionals that can
+// speed up extremely performance-critical spots if the conditional is
+// usually (or rarely) is true.  You use it by replacing
+//     if (x) ...
+// with
+//     if (OSL_LIKELY(x)) ...     // if you think x will usually be true
+// or
+//     if (OSL_UNLIKELY(x)) ...   // if you think x will rarely be true
+// Caveat: Programmers are notoriously bad at guessing this, so it
+// should be used only with thorough benchmarking.
```

Comment:
Perhaps add comment about use in SIMD loops.
// When used in SIMD loops, can influence generation of a test to see if 
// all data lanes are false with a branch to skip the following code block vs. always executing 
// the code block with mask.

Reply:
Can you elaborate on what "can influence" means?

Reply:
@AlexMWells I'm going to merge this as-is, but I can separately modify the comment as you suggested. But just so I'm sure myself and also write the correct comment, can you say more about what you mean? Does using this in SIMD loops make things better or worse, and why?

Reply:
Sure.  when vectorizing and a conditional branch is reached, a compiler will 
1.  Always execute the block with masking (where the mask includes the result of the condition).
2.  Test the conditional on all data lanes to see if they are all false, then jump over the basic block.

For small basic blocks that just do simple math or assignments, it can be better to use approach 1 and just always execute it masked vs. extra instructions to test the condition mask and jmp.  jmp introduces possibility of branch misprediction and may inhibit out of order execution.

For larger or expensive blocks, the cost of the test/jmp is easily amortized and option #2 would be better.  

A compiler will use internal heuristics to make the best decision between #1 and #2 it can.  But heuristics don't cover all cases and may not cover corner cases.

By "influence", I mean that it has been observed that if the compiler is hinted that a branch is unlikely, that it will choose take take option #2.  We make use of this to hint to the compiler that an expensive branch should avoided execution.

Most common use case is for expensive error handling code.  An example would be a check for a result of 0 or NAN, and that if true repeats an algorithm or performs a more expensive version of the algorithm.  The 99% case is that the error won't be happening and we really want to truly skip that block of code.  Usually the heuristic get this, but sometimes the block maybe small but contain an expensive instruction like divide or sqrt.  So even though masked off, it will still take clock cycles.

The inverse is also true, if hinted that a block is "likely", a compiler may choose to test if all lanes are true and execute an "unmasked" version of the block, else jmp to masked version.

I think I am trying to say that  "OSL_LIKELY and OSL_UNLIKELY can influence compiler heuristics on always executing a code block masked vs. emitting a test and jmp to skip a code block entirely."

Reply:
Thanks, that is a very helpful explanation.

Reply:
Would it be correct to say that in loops you hope get auto-vectorized, you should probably be more motivated to use LIKELY/UNLIKELY hints whenever you know it to be true, than you perhaps would in ordinary scalar code where the benefit of this hint is more subtle and hit-or-miss whether it actually helps?

Reply:
In ordinary code, the compiler uses these hints to layout the code favorably to the branch-predictor. But the benefit is definitely more subtle outside of really hot inner loops.

Reply:
I wouldn't go adding them to every conditional statement.  Treat them a fine tuning to override or prevent a compiler heuristic from making the wrong decision.  And yes they can help scalar code as well, however the heuristic being wrong for scalar code is not as detrimental as for vectorized code.

My most common use cases are when I notice unexpected branches always executing, I can use OSL_UNLIKELY to alter that.  Or in cases where I want to make sure the compiler doesn't get it wrong (even if it didn't), maybe a series of if statements that as a developer I know 1 and only 1 will ever executed.    Sometimes it might be self documenting to a reader.

Reply:
Yeah, my experience is that for ordinary code, you need to benchmark, it may or not be helpful in any measurable way. I think Alex seems to be saying that it's much more broadly helpful in vectorized code.